sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -106,7 +106,27 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   final def execute(): RDD[InternalRow] = {
     RDDOperationScope.withScope(sparkContext, nodeName, false, true) {
       prepare()
-      doExecute()
+
+      if (sqlContext.conf.wholeStageEnabled && supportCodeGen
```

Comment:
Here's some code I wrote a while ago that collapses operators by stages when I wanted to refactor the operators to have the local execution mode: https://github.com/apache/spark/compare/master...rxin:local-iter3?expand=1

It is not complete (e.g. doesn't collapse inputs, or handle joins completely), but it could be a good start.